### PR TITLE
[Graphbolt] Add exclude_edges impl and unittest.

### PR DESCRIPTION
**Pull request by czkkkkkk**
## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [ ] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

**Comment by czkkkkkk** [comment]
@dgl-bot

File: python/dgl/graphbolt/impl/sampled_subgraph_impl.py [hunk 1]
```diff
@@ -94,5 +151,39 @@ def exclude_edges(
...
         The sampled subgraph with the excluded edges.
     """
-    # TODO(zhenkun): Implement this.
-    raise NotImplementedError
+    assert isinstance(subgraph.node_pairs, tuple) == isinstance(edges, tuple), (
+        "The sampled subgraph and the edges to exclude should be both "
+        "homogeneous or both heterogeneous."
+    )
+    # Three steps to exclude edges:
+    # 1. Convert the node pairs to the reverse ids.
```
Comment (frozenbugs) [inline python/dgl/graphbolt/impl/sampled_subgraph_impl.py:159]:
```suggestion
    # 1. Convert the node pairs to the original ids if they are compacted.
```

File: python/dgl/graphbolt/impl/sampled_subgraph_impl.py [hunk 2]
```diff
@@ -68,6 +68,63 @@ def __post_init__(self):
...
+    _, src = torch.unique(src, return_inverse=True)
+    _, dst = torch.unique(dst, return_inverse=True)
+    lhs_edges = (src[: lhs_edges[0].numel()], dst[: lhs_edges[0].numel()])
+    rhs_edges = (src[lhs_edges[0].numel() :], dst[lhs_edges[0].numel() :])
+    return lhs_edges, rhs_edges
+
+
+def _exclude_homo_edges(lhs_edges, rhs_edges):
+    lhs_edges, rhs_edges = _relabel_homo_edges(lhs_edges, rhs_edges)
+    # Compact edges to integers to use torch.isin to exclude edges
```
Comment (frozenbugs) [inline python/dgl/graphbolt/impl/sampled_subgraph_impl.py:94]:
```suggestion
    # Compact edges to integers to use torch.isin to exclude edges.
```

File: python/dgl/graphbolt/impl/sampled_subgraph_impl.py [hunk 3]
```diff
@@ -68,6 +68,63 @@ def __post_init__(self):
...
+    """Slice the subgraph according to the index."""
+
+    def _index_select(obj, index):
+        if obj is None:
+            return None
+        if isinstance(obj, torch.Tensor):
+            return obj[index]
+        if isinstance(obj, tuple):
+            return tuple(_index_select(v, index) for v in obj)
+        # Handle the case when obj is a dictionary
```
Comment (frozenbugs) [inline python/dgl/graphbolt/impl/sampled_subgraph_impl.py:112]: ditto

File: python/dgl/graphbolt/impl/sampled_subgraph_impl.py [hunk 4]
```diff
@@ -68,6 +68,63 @@ def __post_init__(self):
...
+    dst = torch.cat([lhs_edges[1], rhs_edges[1]])
+    _, src = torch.unique(src, return_inverse=True)
+    _, dst = torch.unique(dst, return_inverse=True)
+    lhs_edges = (src[: lhs_edges[0].numel()], dst[: lhs_edges[0].numel()])
+    rhs_edges = (src[lhs_edges[0].numel() :], dst[lhs_edges[0].numel() :])
+    return lhs_edges, rhs_edges
+
+
+def _exclude_homo_edges(lhs_edges, rhs_edges):
+    lhs_edges, rhs_edges = _relabel_homo_edges(lhs_edges, rhs_edges)
```
Comment (frozenbugs) [inline python/dgl/graphbolt/impl/sampled_subgraph_impl.py:93]: Why do we need to relabel before we exclude? It is a little bit confusing.

File: python/dgl/graphbolt/impl/sampled_subgraph_impl.py [hunk 5]
```diff
@@ -68,6 +68,63 @@ def __post_init__(self):
...
+            return tuple(_index_select(v, index) for v in obj)
+        # Handle the case when obj is a dictionary
+        assert isinstance(obj, dict)
+        assert isinstance(index, dict)
+        ret = {}
+        for k, v in obj.items():
+            ret[k] = _index_select(v, index[k])
+        return ret
+
+    return SampledSubgraphImpl(
```
Comment (frozenbugs) [inline python/dgl/graphbolt/impl/sampled_subgraph_impl.py:120]: Is the returned SampledSubgraphImpl compacted or not or depends on the input SampledSubgraphImpl?

File: tests/python/pytorch/graphbolt/impl/test_sampled_subgraph_impl.py [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+            assert key in rhs
+            _assert_container_equal(value, rhs[key])
+
+
+@pytest.mark.parametrize("is_reverse", [True, False])
+def test_exclude_edges_homo(is_reverse):
+    node_pairs = (torch.tensor([0, 2, 3]), torch.tensor([1, 4, 2]))
+    if is_reverse:
+        reverse_row_node_ids = torch.tensor([10, 15, 11, 24, 9])
+        reverse_column_node_ids = torch.tensor([10, 15, 11, 24, 9])
```
Comment (frozenbugs) [inline tests/python/pytorch/graphbolt/impl/test_sampled_subgraph_impl.py:31]: @peizhou001 do we compact row and column respectively or altogether?

File: tests/python/pytorch/graphbolt/impl/test_sampled_subgraph_impl.py [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+    subgraph = SampledSubgraphImpl(
+        node_pairs,
+        reverse_column_node_ids,
+        reverse_row_node_ids,
+        reverse_edge_ids,
+    )
+    if is_reverse:
+        excluded_edges = (torch.tensor([11]), torch.tensor([9]))
+    else:
+        excluded_edges = (torch.tensor([2]), torch.tensor([4]))
```
Comment (frozenbugs) [inline tests/python/pytorch/graphbolt/impl/test_sampled_subgraph_impl.py:44]: by any chance user will use compacted edge id?

File: python/dgl/graphbolt/impl/sampled_subgraph_impl.py [hunk 8]
```diff
@@ -68,6 +68,63 @@ def __post_init__(self):
...
+def _to_reverse_ids(node_pair, reverse_row_node_ids, reverse_column_node_ids):
+    u, v = node_pair
+    if reverse_row_node_ids is not None:
+        u = reverse_row_node_ids[u]
+    if reverse_column_node_ids is not None:
+        v = reverse_column_node_ids[v]
+    return (u, v)
+
+
+def _relabel_homo_edges(lhs_edges, rhs_edges):
```
Comment (BarclayII) [inline python/dgl/graphbolt/impl/sampled_subgraph_impl.py:80]: Why do we have two sets of edges during relabeling?

Comment (peizhou001) [reply python/dgl/graphbolt/impl/sampled_subgraph_impl.py:80]: According to previous discussion, some sampling algorithm natually have column compressed from a view of "columnwise pick", but row is not compressed.

File: tests/python/pytorch/graphbolt/impl/test_sampled_subgraph_impl.py (hunk 6 above)
Comment (peizhou001) [reply tests/python/pytorch/graphbolt/impl/test_sampled_subgraph_impl.py:31]: We compact them together. But according to our previous discussion, row and column may not be both compressed in some algorithm.

File: python/dgl/graphbolt/impl/sampled_subgraph_impl.py [hunk 9]
```diff
@@ -94,5 +151,39 @@ def exclude_edges(
     SampledSubgraphImpl
         The sampled subgraph with the excluded edges.
```
Comment (peizhou001) [inline python/dgl/graphbolt/impl/sampled_subgraph_impl.py:152]: without?

File: python/dgl/graphbolt/impl/sampled_subgraph_impl.py [hunk 10]
```diff
@@ -94,5 +151,39 @@ def exclude_edges(
...
     """
-    # TODO(zhenkun): Implement this.
-    raise NotImplementedError
+    assert isinstance(subgraph.node_pairs, tuple) == isinstance(edges, tuple), (
+        "The sampled subgraph and the edges to exclude should be both "
+        "homogeneous or both heterogeneous."
```
Comment (peizhou001) [inline python/dgl/graphbolt/impl/sampled_subgraph_impl.py:156]: Remove the second both?

Comment (peizhou001) [inline python/dgl/graphbolt/impl/sampled_subgraph_impl.py:156]: Remove the second both?

File: python/dgl/graphbolt/impl/sampled_subgraph_impl.py [hunk 11]
```diff
@@ -68,6 +68,63 @@ def __post_init__(self):
...
+    lhs_edges = (src[: lhs_edges[0].numel()], dst[: lhs_edges[0].numel()])
+    rhs_edges = (src[lhs_edges[0].numel() :], dst[lhs_edges[0].numel() :])
+    return lhs_edges, rhs_edges
+
+
+def _exclude_homo_edges(lhs_edges, rhs_edges):
+    lhs_edges, rhs_edges = _relabel_homo_edges(lhs_edges, rhs_edges)
+    # Compact edges to integers to use torch.isin to exclude edges
+    dst_range = lhs_edges[0].numel() + rhs_edges[0].numel()
+    lhs_val = lhs_edges[0] * dst_range + lhs_edges[1]
```
Comment (peizhou001) [inline python/dgl/graphbolt/impl/sampled_subgraph_impl.py:96]: Can I understand first element in tuple as high number and the second as low number with  `dst_range` as radix?
If this is the case, will it cause error because both` lhs_edges[1]` and `rhs_edges[1]` can be larger than `dst_range`?

File: python/dgl/graphbolt/impl/sampled_subgraph_impl.py [hunk 12]
```diff
@@ -68,6 +68,63 @@ def __post_init__(self):
...
+    rhs_val = rhs_edges[0] * dst_range + rhs_edges[1]
+    mask = ~torch.isin(lhs_val, rhs_val)
+    return torch.nonzero(mask, as_tuple=True)[0]
```
Comment (peizhou001) [inline python/dgl/graphbolt/impl/sampled_subgraph_impl.py:99]: Another choice it to return mask directly and use `[mask]` for filter. Both methonds all okay for me, you can choose one you perfer.

File: python/dgl/graphbolt/impl/sampled_subgraph_impl.py (hunk 4 above)
Comment (peizhou001) [reply python/dgl/graphbolt/impl/sampled_subgraph_impl.py:93]: I'm also a little confused as both ids are already converted to non-compacted format. And `unique `should be a relatively costly op.

File: python/dgl/graphbolt/impl/sampled_subgraph_impl.py [hunk 13]
```diff
@@ -68,6 +68,63 @@ def __post_init__(self):
             ), "Nodes in pairs should be of type torch.Tensor."
 
 
+def _to_reverse_ids(node_pair, reverse_row_node_ids, reverse_column_node_ids):
```
Comment (peizhou001) [inline python/dgl/graphbolt/impl/sampled_subgraph_impl.py:71]: ` reverse_row_node_ids` is not the compacted counterpart of  `node_pair[0]`, same for `reverse_column_node_ids`. For example:

```
origin_node_pair = ([100, 50, 50, 100], ...)
after compact 
node_pair = ([1, 0, 0, 1], ...)
reverese_row_node_ids = [50, 100], means 50 -> 0 and 100 -> 1.
```

**Review by peizhou001 — CHANGES_REQUESTED** [review]

File: python/dgl/graphbolt/impl/sampled_subgraph_impl.py (hunk 13 above)
Comment (czkkkkkk) [reply python/dgl/graphbolt/impl/sampled_subgraph_impl.py:71]: This is also my understanding. The code below is to get the `origin_node_pair` of your example. It works because `reverse_row_node_ids[node_pair] = [100, 50, 50, 100]`.

File: python/dgl/graphbolt/impl/sampled_subgraph_impl.py (hunk 8 above)
Comment (czkkkkkk) [reply python/dgl/graphbolt/impl/sampled_subgraph_impl.py:80]: This to to map the node ids of two set of edges to a consecutive range from 0. I concatenate them in this function.

File: python/dgl/graphbolt/impl/sampled_subgraph_impl.py (hunk 4 above)
Comment (czkkkkkk) [reply python/dgl/graphbolt/impl/sampled_subgraph_impl.py:93]: This is to convert each edge to a single value to apply `torch.isin`. Another solution is to compact an edge `(u, v)` to `(u << 32 + v)`. But it does not work if some node id exceed `2^32`. Therefore I further compact the nodes ids of all edges to make their range smaller.

File: python/dgl/graphbolt/impl/sampled_subgraph_impl.py (hunk 11 above)
Comment (czkkkkkk) [reply python/dgl/graphbolt/impl/sampled_subgraph_impl.py:96]: They would not be larger than `lhs_edges[0].numel() + rhs_edges[0].numel()` since they are compacted.

File: python/dgl/graphbolt/impl/sampled_subgraph_impl.py (hunk 12 above)
Comment (czkkkkkk) [reply python/dgl/graphbolt/impl/sampled_subgraph_impl.py:99]: Yeah. I think it does not affect the performance.

File: python/dgl/graphbolt/impl/sampled_subgraph_impl.py (hunk 5 above)
Comment (czkkkkkk) [reply python/dgl/graphbolt/impl/sampled_subgraph_impl.py:120]: It depends on the input subgraph. I will add explanation about this.

File: tests/python/pytorch/graphbolt/impl/test_sampled_subgraph_impl.py (hunk 6 above)
Comment (czkkkkkk) [reply tests/python/pytorch/graphbolt/impl/test_sampled_subgraph_impl.py:31]: My implementation does not assume they are compacted together. I will add tests on the cases of single compaction.

File: tests/python/pytorch/graphbolt/impl/test_sampled_subgraph_impl.py (hunk 7 above)
Comment (czkkkkkk) [reply tests/python/pytorch/graphbolt/impl/test_sampled_subgraph_impl.py:44]: Yeah. This should be removed.

File: python/dgl/graphbolt/impl/sampled_subgraph_impl.py (hunk 8 above)
Comment (BarclayII) [reply python/dgl/graphbolt/impl/sampled_subgraph_impl.py:80]: I see.  I would suggest (1) rename `lhs_edges` and `rhs_edges` to more informative names like `node_id_pairs`, `node_id_pairs_to_exclude`.

File: python/dgl/graphbolt/impl/sampled_subgraph_impl.py [hunk 14]
```diff
@@ -68,6 +68,63 @@ def __post_init__(self):
...
+
+def _relabel_homo_edges(lhs_edges, rhs_edges):
+    """Relabel the source nodes and destination nodes independently for two set
+    of edges."""
+    src = torch.cat([lhs_edges[0], rhs_edges[0]])
+    dst = torch.cat([lhs_edges[1], rhs_edges[1]])
+    _, src = torch.unique(src, return_inverse=True)
+    _, dst = torch.unique(dst, return_inverse=True)
+    lhs_edges = (src[: lhs_edges[0].numel()], dst[: lhs_edges[0].numel()])
+    rhs_edges = (src[lhs_edges[0].numel() :], dst[lhs_edges[0].numel() :])
```
Comment (BarclayII) [inline python/dgl/graphbolt/impl/sampled_subgraph_impl.py:88]: Try to avoid reusing variables for different purposes, e.g. in L85-86, `src` could be something like `src_relabeled`.

File: tests/python/pytorch/graphbolt/impl/test_sampled_subgraph_impl.py (hunk 7 above)
Comment (czkkkkkk) [reply tests/python/pytorch/graphbolt/impl/test_sampled_subgraph_impl.py]: I found the `if` branch cannot be removed in the tests. For the case of `is_reverse == False`, it means the subgraph is not compacted and the edge `(2, 4)` means an edge between original ids.

File: python/dgl/graphbolt/impl/sampled_subgraph_impl.py (hunk 8 above)
Comment (czkkkkkk) [reply python/dgl/graphbolt/impl/sampled_subgraph_impl.py]: Good point. I use `edges` and `edges_to_exclude` now.

File: python/dgl/graphbolt/impl/sampled_subgraph_impl.py (hunk 14 above)
Comment (czkkkkkk) [reply python/dgl/graphbolt/impl/sampled_subgraph_impl.py]: I rewrote this function to make it only relabel two arrays instead of two edge sets.

**Review by frozenbugs — APPROVED** [review]

File: tests/python/pytorch/graphbolt/impl/test_sampled_subgraph_impl.py [hunk 15]
```diff
@@ -0,0 +1,136 @@
...
+    reverse_edge_ids = torch.Tensor([5, 9, 10])
+    subgraph = SampledSubgraphImpl(
+        node_pairs,
+        reverse_column_node_ids,
+        reverse_row_node_ids,
+        reverse_edge_ids,
+    )
+    src_to_exclude = torch.tensor([11]) if reverse_row else torch.tensor([2])
+    dst_to_exclude = torch.tensor([9]) if reverse_column else torch.tensor([4])
+    excluded_edges = (src_to_exclude, dst_to_exclude)
```
Comment (frozenbugs) [inline tests/python/pytorch/graphbolt/impl/test_sampled_subgraph_impl.py:48]: For excluded_edges, should them all be original hetero id?

File: tests/python/pytorch/graphbolt/impl/test_sampled_subgraph_impl.py [hunk 16]
```diff
@@ -0,0 +1,136 @@
...
+    result = exclude_edges(subgraph, excluded_edges)
```
Comment (frozenbugs) [inline tests/python/pytorch/graphbolt/impl/test_sampled_subgraph_impl.py:49]:
```suggestion
    result = exclude_edges(subgraph, edges_to_exclude)
```
